README
MAIN PURPOSE

Ugh is HTTP server, for which it's very simple to write module, which goes to
different backends simultaneously, then does something with their response,
then goes again to some other backends et cetera.

This is achieved by using coroutines mechanism. Each request is handled in a
separate coroutine and the main coroutine is for server's event loop. That's
why you could ask for several subrequests, then just call ugh_subreq_wait() to
wait for their results and then, just after ugh_subreq_wait() is returned, you
could use all the results from subrequest, like it was synchronous code.

You could do simultaneous subrequests even without writing your own module, but
ugh is much more flexible and powerful if you're using its API, not only
configuration.

CONFIGURATION

Configuration of ugh tries to be like nginx. Actually it's much simpler, cause
ugh doesn't support most of nginx's features, and does only its specific kind
of job (simultaneous subrequests with non-trivial logic of handling their
responses).

Example of config file is included in repository and is self-explanatory.
Possible directives include:

error_log file [debug | info | notice | warn | error | crit | alert | emerg];
  Log errors into file (with optional severity argument).

listen [address:]port;
  What address:port to listen to.

resolver host;
  Use specific DNS-server to resolve hosts of subrequest URLs.

import module_name [path/to/module.so];
  Import third-party module from .so-library. By default ugh searches inside
  PREFIX/lib/ugh/ for library named libugh_module_name.so.

map { ... }
  Just like nginx's map (doesn't support hostnames flag, though).

upstream { ... }
  Just like nginx's upstream, but doesn't support all server's options and
  backup is not an option, but is different directive. See config example.
  Supported options are max_fails and fail_timeout, but there's an importante
  difference between nginx and ugh in how they work:
    a) in ugh max_fails / fail_timeout work even for one-server upstream;
    b) max_fails is counted as consequtive failures, not as failures for
       fail_timeout period of time.

proxy_pass url;
  Just like nginx's proxy_pass, but do not use scheme in its argument.

proxy_nowait on | off;
  Wait or don't wait for response of the latest previous proxy_pass in config.

proxy_recv_timeout timeout;
  Just like nginx's one, affects only latest previous proxy_pass in config.

proxy_next_upstream type;
  Just like nginx's one, affects only latest previous proxy_pass in config.

return data;
  Return some data with 200 OK.

One more note about configuration is that you could use variables and templates
(strings with variables inside) like in nginx. There are some built-in
variables. You can make your own with map directive. Directives return and
proxy_pass support variables in their arguments. The list of built-in
variables:

$arg_... - GET-argument with "..." name.
$http_... - HTTP header with "..." name.
$cookie_... - HTTP cookie with "..." name.
$hash_... - hash of value of variable with "..." name.
$c0_... - first char of value of variable with "..." name.
$cl_... - last char of value of variable with "..." name.

$var_... - value of custom variable with "..." name (can be set by module, this
might be removed in later versions in favor of more flexible mechanism).

COMPILE AND INSTALL

Any Unix/Linux system:
$ cmake .
$ make
$ make install

Debial/Ubuntu package:
$ dpkg-buildpackage -rfakeroot

WRITING MODULES

See module example in repository.



debian/changelog
ugh (0.1.15-1) unstable; urgency=medium

  * Added support for HTTP statuses 417 and 451

 -- Ruslan Osmanov <rrosmanov@gmail.com>  Mon, 13 Sep 2019 18:13:00 +0300

ugh (0.1.14-1) unstable; urgency=medium

  * supported templates of unlimited length

 -- Sergey Bochenkov <ba4an@yandex.ru>  Thu, 13 Sep 2018 15:45:03 +0700

ugh (0.1.13-1) unstable; urgency=medium

  * Fixed dependency to library 'm' for Debian 9

 -- Konstantin Akimov <knstqq@gmail.com>  Wed, 06 Dec 2017 19:45:03 +0700

ugh (0.1.12-1) unstable; urgency=medium

  * fixed subreq headers reading

 -- Sergey Bochenkov <ba4an@yandex.ru>  Tue, 04 Apr 2017 12:38:58 +0700

ugh (0.1.11-1) unstable; urgency=medium

  * support setting out headers for subrequest

 -- Sergey Bochenkov <ba4an@yandex.ru>  Tue, 20 Dec 2016 00:38:58 +0700

ugh (0.1.10-10) unstable; urgency=medium

  * clear next_upstream value if user sets it from config

 -- Sergey Bochenkov <ba4an@yandex.ru>  Fri, 16 Dec 2016 17:38:58 +0700

ugh (0.1.10-9) unstable; urgency=medium

  * fixed bug in ugh_subreq_set_method

 -- Sergey Bochenkov <ba4an@yandex.ru>  Wed, 30 Nov 2016 16:08:58 +0700

ugh (0.1.10-8) unstable; urgency=medium

  * version bump

 -- Sergey Bochenkov <ba4an@yandex.ru>  Thu, 24 Nov 2016 19:18:58 +0700

ugh (0.1.10-7) unstable; urgency=low

  * support 204 and 304 responses properly

 -- Sergey Bochenkov <ba4an@yandex.ru>  Mon, 08 Nov 2016 14:54:05 +0700

ugh (0.1.10-6) unstable; urgency=low

  * support failover for close-after-body responses

 -- Sergey Bochenkov <ba4an@yandex.ru>  Mon, 24 Oct 2016 17:54:05 +0700

ugh (0.1.10-5) unstable; urgency=low

  * always write correct content-length for POST subrequests

 -- Sergey Bochenkov <ba4an@yandex.ru>  Mon, 16 May 2016 15:23:05 +0700

ugh (0.1.10-4) unstable; urgency=low

  * fixed max_fails / fail_timeout logic

 -- bachan <bachan@farm.itim.vn>  Mon, 16 Oct 2015 10:57:12 +0700

ugh (0.1.10-3) unstable; urgency=low

  * install modules into lib

 -- bachan <bachan@farm.itim.vn>  Mon, 05 Oct 2015 10:57:12 +0700

ugh (0.1.10-2) unstable; urgency=low

  * fixed max_fails feature

 -- bachan <bachan@farm.itim.vn>  Thu, 01 Oct 2015 13:46:30 +0700

ugh (0.1.10-1) unstable; urgency=low

  * support max_fails / fail_timeout (meaning is different from nginx)

 -- bachan <bachan@farm.itim.vn>  Tue, 22 Sep 2015 12:38:02 +0700

ugh (0.1.9-1) unstable; urgency=low

  * fixed resolver cache
  * added cookie_va function

 -- bachan <bachan@farm.itim.vn>  Thu, 06 Aug 2015 19:14:18 +0700

ugh (0.1.8-1) unstable; urgency=low

  * added connection_time field

 -- bachan <bachan@farm.itim.vn>  Mon, 15 Dec 2014 14:09:29 +0700

ugh (0.1.7-5) unstable; urgency=high

  * reinit http parser state during failover

 -- bachan <bachan@farm.itim.vn>  Thu,  4 Sep 2014 18:10:00 +0700

ugh (0.1.7-4) unstable; urgency=high

  * fixed failover in case of connections prematurely closed by upstream

 -- bachan <bachan@farm.itim.vn>  Tue, 19 Aug 2014 14:44:00 +0700

ugh (0.1.7-3) unstable; urgency=high

  * fixed default next_upstream setting

 -- bachan <bachan@farm.itim.vn>  Mon, 14 Jul 2014 17:21:00 +0700

ugh (0.1.7-2) unstable; urgency=low

  * renamed some macros to avoid compilation errors with some libraries

 -- bachan <bachan@farm.itim.vn>  Wed,  7 Jul 2014 12:50:27 +0700

ugh (0.1.7-1) unstable; urgency=low

  * add separate ft_type for connect timeout

 -- bachan <bachan@farm.itim.vn>  Wed, 19 Mar 2014 14:50:27 +0700

ugh (0.1.6-4) unstable; urgency=low

  * fix pid-file race condition

 -- bachan <bachan@farm.itim.vn>  Wed, 19 Mar 2014 11:50:27 +0700

ugh (0.1.6-3) unstable; urgency=low

  * log string about connect timeout

 -- bachan <bachan@farm.itim.vn>  Mon, 17 Mar 2014 22:13:27 +0700

ugh (0.1.6-2) unstable; urgency=low

  * log string about connect timeout

 -- bachan <bachan@farm.itim.vn>  Mon, 17 Mar 2014 22:03:27 +0700

ugh (0.1.6-1) unstable; urgency=low

  * connection timeouts support

 -- bachan <bachan@farm.itim.vn>  Mon, 17 Mar 2014 18:03:27 +0700

ugh (0.1.5-1) unstable; urgency=low

  * multiple set-cookie header support

 -- bachan <ba4an@yandex.ru>  Wed, 13 Nov 2013 16:30:20 +0700

ugh (0.1.4-1) unstable; urgency=low

  * made UGH_TEMPLATE_MAX_CHUNKS bigger

 -- bachan <ba4an@yandex.ru>  Mon, 12 Oct 2013 14:30:20 +0700

ugh (0.1.3-1) unstable; urgency=low

  * fixed possibility of buffer overflow in subrequest

 -- bachan <ba4an@yandex.ru>  Thu, 03 Oct 2013 17:30:20 +0700

ugh (0.1.2-2) unstable; urgency=low

  * aux_urlenc, aux_urldec

 -- bachan <ba4an@yandex.ru>  Fri, 28 Aug 2013 18:40:20 +0700

ugh (0.1.2-1) unstable; urgency=low

  * choose random directive in upstream

 -- bachan <ba4an@yandex.ru>  Fri, 23 Aug 2013 17:53:20 +0700

ugh (0.1.1-7) unstable; urgency=low

  * ugh_client_setvar_va

 -- bachan <ba4an@yandex.ru>  Fri, 23 Aug 2013 17:02:20 +0700

ugh (0.1.1-6) unstable; urgency=low

  * ugh_make_command_double

 -- bachan <ba4an@yandex.ru>  Fri, 23 Aug 2013 15:02:20 +0700

ugh (0.1.1-5) unstable; urgency=low

  * ugh_client_setvar_cp

 -- bachan <ba4an@yandex.ru>  Tue, 21 Aug 2013 14:02:20 +0700

ugh (0.1.1-4) unstable; urgency=low

  * bugfix in resolv_conf parser

 -- bachan <ba4an@yandex.ru>  Tue, 20 Aug 2013 18:20:20 +0700

ugh (0.1.1-3) unstable; urgency=low

  * fix clearing full timeout when going to the next upstream

 -- bachan <ba4an@yandex.ru>  Tue, 30 Jul 2013 20:57:20 +0700

ugh (0.1.1-2) unstable; urgency=low

  * recv_timeout fixes, version fix

 -- bachan <ba4an@yandex.ru>  Mon, 29 Jul 2013 18:20:20 +0700

ugh (0.0.10-1) unstable; urgency=low

  * changed module API

 -- bachan <bachan@sample-sandbox1g.dev.itim.vn>  Mon, 29 Jul 2013 17:33:20 +0700

ugh (0.0.9-1) unstable; urgency=low

  * lots of fixes around channel stuff

 -- bachan <bachan@sample-sandbox1g.dev.itim.vn>  Tue, 23 Jul 2013 11:36:28 +0700

ugh (0.0.7-2) unstable; urgency=low

  * don't remove debug symbols

 -- bachan <bachan@sample-sandbox1g.dev.itim.vn>  Wed, 10 Jul 2013 18:26:24 +0700


ugh/ugh.h
#ifndef __UGH_H__
#define __UGH_H__

#include <ev.h>
#include <Judy.h>
#include <stdarg.h>
#include <stddef.h>
#include <stdlib.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <netinet/in.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/uio.h>
#include "aux/buffer.h"
#include "aux/config.h"
#include "aux/hashes.h"
#include "aux/daemon.h"
#include "aux/logger.h"
#include "aux/memory.h"
#include "aux/random.h"
#include "aux/socket.h"
#include "aux/string.h"
#if 0
#ifndef CORO_UCONTEXT
#define CORO_UCONTEXT 1
#endif
#include "coro/coro.h"
#endif
#include "coro_ucontext/coro_ucontext.h"
#include <ugh/autoconf.h>
#include "config.h"
#include "module.h"
#include "resolver.h"

#ifdef __cplusplus
extern "C" {
#endif

/* ### ugh_server ########################################################## */

typedef struct ugh_server
	ugh_server_t;

struct ugh_server
{
	ugh_config_t *cfg;
	ugh_resolver_t *resolver;
	struct sockaddr_in addr;
	ev_io wev_accept;
	void *channels_hash; /* JudyL (channel_id -> ugh_channel_t *) */
};

int ugh_server_listen(ugh_server_t *s, ugh_config_t *cfg, ugh_resolver_t *resolver);
int ugh_server_enough(ugh_server_t *s);

/* ### ugh_daemon ########################################################## */

typedef struct ugh_daemon
	ugh_daemon_t;

struct ugh_daemon
{
	ugh_config_t cfg;
	ugh_server_t srv;
	ugh_resolver_t resolver;

	ev_timer wev_silent;
};

int ugh_daemon_exec(const char *cfg_filename, unsigned daemon);

extern struct ev_loop *loop;

/* ### ugh_client ########################################################## */

#define UGH_OK 0
#define UGH_ERROR -1
#define UGH_AGAIN -2

#define UGH_HTTP_GET    0
#define UGH_HTTP_HEAD   1
#define UGH_HTTP_POST   2
#define UGH_HTTP_PUT    3
#define UGH_HTTP_DELETE 4
#define UGH_HTTP_MAX    5

#define UGH_HTTP_VERSION_0_9 0
#define UGH_HTTP_VERSION_1_0 1
#define UGH_HTTP_VERSION_1_1 2
#define UGH_HTTP_VERSION_MAX 3

extern const char *ugh_method_string [UGH_HTTP_MAX];
extern const char *ugh_version_string [UGH_HTTP_VERSION_MAX];

#define UGH_HEADER_BUF (32768) /* XXX this is temporary upper header buffer size limit to simplify things */
#define UGH_SUBREQ_BUF (32768 * 4) /* XXX make this dynamically allocated also */
#define UGH_CHUNKS_BUF (32768 * 4) /* start buffer size for chunked response */
#define UGH_CORO_STACK (32768 * 16)

typedef struct ugh_client
	ugh_client_t;

typedef struct ugh_subreq
	ugh_subreq_t;

typedef struct ugh_channel
	ugh_channel_t;

struct ugh_client
{
	ugh_server_t *s;
	struct sockaddr_in addr;

	ev_io wev_recv;
	ev_io wev_send;
	ev_timer wev_timeout;

	aux_pool_t *pool;

	strt buf_recv; /* UGH_HEADER_BUF */
	strt buf_send;

	/* request */

	char *key_b;
	char *key_e;
	char *val_b;
	char *val_e;

	char *request_beg;
	char *headers_beg;
	char *request_end;

	unsigned char state;
	unsigned char method;
	unsigned char version;

	strt uri;
	strt args;
	void *args_hash;
	void *headers_hash;
	void *body_hash; /* posted args (application/x-www-form-urlencoded) */
#if 1
	void *vars_hash;
#endif

	strt body;
	size_t content_length;

	void *subreqs_hash; /* Judy1 (ugh_subreq_t *) */

	/* response */

	void *headers_out_hash;
	void *cookies_out_hash;
	strp bufs;
	size_t bufs_sumlen;

	struct iovec *iov;
	size_t iov_num;

#if 1
	char *stack;
	coro_context ctx;
#endif

	unsigned wait; /* num of subrequests to wait for */
};

int ugh_client_add(ugh_server_t *s, int sd, struct sockaddr_in *addr);
int ugh_client_del(ugh_client_t *c);

int ugh_client_add_subreq(ugh_client_t *c, ugh_subreq_t *r);
int ugh_client_del_subreqs(ugh_client_t *c);

int ugh_client_send(ugh_client_t *c, int status);

strp ugh_client_getarg_nt(ugh_client_t *c, const char *data);
strp ugh_client_getarg(ugh_client_t *c, const char *data, size_t size);
strp ugh_client_setarg(ugh_client_t *c, const char *data, size_t size, char *value_data, size_t value_size);

strp ugh_client_body_getarg_nt(ugh_client_t *c, const char *data);
strp ugh_client_body_getarg(ugh_client_t *c, const char *data, size_t size);
strp ugh_client_body_setarg(ugh_client_t *c, const char *data, size_t size, char *value_data, size_t value_size);

typedef struct ugh_header
	ugh_header_t;

struct ugh_header
{
	strt key;
	strt value;
};

ugh_header_t *ugh_client_header_get_nt(ugh_client_t *c, const char *data);
ugh_header_t *ugh_client_header_get(ugh_client_t *c, const char *data, size_t size);
ugh_header_t *ugh_client_header_set(ugh_client_t *c, const char *data, size_t size, char *value_data, size_t value_size);
ugh_header_t *ugh_client_header_out_set(ugh_client_t *c, const char *data, size_t size, char *value_data, size_t value_size);
#define ugh_client_header_out_set_nt(c, data, value_data) ugh_client_header_out_set((c), (data), strlen(data), (char *) (value_data), strlen(value_data))

strp ugh_client_cookie_get(ugh_client_t *c, const char *name, size_t size);
#define ugh_client_cookie_get_nt(c, name) ugh_client_cookie_get((c), (name), strlen(name))
strp ugh_client_cookie_out_set(ugh_client_t *c, char *data, size_t size);
strp ugh_client_cookie_out_set_va(ugh_client_t *c, const char *fmt, ...) AUX_FORMAT(printf, 2, 3);

strp ugh_client_getvar(ugh_client_t *c, const char *data, size_t size);
strp ugh_client_getvar_nt(ugh_client_t *c, const char *data);
strp ugh_client_setvar(ugh_client_t *c, const char *data, size_t size, char *value_data, size_t value_size);
strp ugh_client_setvar_nt(ugh_client_t *c, const char *data, char *value_data, size_t value_size);
strp ugh_client_setvar_cp(ugh_client_t *c, const char *data, char *value_data, size_t value_size);
strp ugh_client_setvar_va(ugh_client_t *c, const char *data, const char *fmt, ...) AUX_FORMAT(printf, 3, 4);

/* ### url ### */

typedef struct ugh_url
	ugh_url_t;

struct ugh_url
{
	strt host;
	strt port;
	strt uri;
	strt args;
};

int ugh_parser_url(ugh_url_t *u, char *data, size_t size);

/* ### upstream ### */

#define UGH_MAX_UPSTREAM_ELEMENTS 1024

typedef struct ugh_upstream_server
	ugh_upstream_server_t;

struct ugh_upstream_server
{
	strt host;
	in_port_t port;

	unsigned max_fails;
	unsigned fails;

	ev_tstamp fail_timeout;
	ev_tstamp fail_start;
};

typedef struct ugh_upstream
	ugh_upstream_t;

struct ugh_upstream
{
	ugh_upstream_server_t values [UGH_MAX_UPSTREAM_ELEMENTS];
	size_t values_size;
	size_t values_curr;

	ugh_upstream_server_t backup_values [UGH_MAX_UPSTREAM_ELEMENTS];
	size_t backup_values_size;
	size_t backup_values_curr;
};

ugh_upstream_t *ugh_upstream_add(ugh_config_t *cfg, const char *name, size_t size);
ugh_upstream_t *ugh_upstream_get(ugh_config_t *cfg, const char *name, size_t size);

/* ### subreq ### */

typedef int (*ugh_subreq_handle_fp)(ugh_subreq_t *r, char *data, size_t size);

#define UGH_TIMEOUT_ONCE 0 /* TODO this is deprecated, we should just have 4 separate timeouts: full, recv, send and connect */
#define UGH_TIMEOUT_FULL 1

#define UGH_RESPONSE_CHUNKED -1
#define UGH_RESPONSE_CLOSE_AFTER_BODY -2

struct ugh_subreq
{
	ugh_client_t *c;
	struct sockaddr_in addr;

	ev_io wev_recv;
	ev_io wev_send;
	ev_io wev_connect;

	ev_timer wev_timeout;
	ev_tstamp timeout;
	int timeout_type;

	ev_timer wev_timeout_connect;
	ev_tstamp timeout_connect;

	/* send */

	ugh_url_t u;
	strt request_body;
	unsigned char method;
	void *headers_out_hash; /* headers to send to backend */
	aux_buffer_t b_send;

	/* recv */

	char *key_b;
	char *key_e;
	char *val_b;
	char *val_e;

	char *request_beg;
	char *headers_beg;
	char *request_end;

	unsigned char state;
	unsigned char version;
	unsigned status;

	void *headers_hash; /* headers received from backend */

	char *buf_recv_data;
	strt buf_recv; /* UGH_SUBREQ_BUF */
	strt body;

	size_t content_length; /* UGH_RESPONSE_* may be encoded here */

	ev_tstamp connection_time;
	ev_tstamp response_time;
	uint32_t ft_type;

	/* chunks */

	char *chunk_start;
	size_t chunk_size;
	size_t chunk_body_size;

	/* done */

	int flags;
	ugh_subreq_handle_fp handle;

	/* tmp */

	ugh_resolver_ctx_t *resolver_ctx;

	/* tmp */

	ugh_upstream_t *upstream;
	unsigned upstream_current;
	unsigned upstream_tries;

	/* push */

	ugh_channel_t *ch;
	unsigned tag; /* you can get this later from the message, so it'll help you to distinguish different sources */
};

#define UGH_SUBREQ_WAIT 1
#define UGH_SUBREQ_PUSH 2

ugh_subreq_t *ugh_subreq_add(ugh_client_t *c, char *url, size_t size, int flags);
int ugh_subreq_set_method(ugh_subreq_t *r, unsigned char method);
int ugh_subreq_set_body(ugh_subreq_t *r, char *body, size_t body_size);
int ugh_subreq_set_timeout(ugh_subreq_t *r, ev_tstamp timeout, int timeout_type);
int ugh_subreq_set_timeout_connect(ugh_subreq_t *r, ev_tstamp timeout);

/* NOTE: tag helps you to distinguish different subrequests, etag helps you to
 * sort received messages, but if you just want to get them as-is (unsorted,
 * use etag=0)  */
int ugh_subreq_set_channel(ugh_subreq_t *r, ugh_channel_t *ch, unsigned tag);

int ugh_subreq_run(ugh_subreq_t *r);
int ugh_subreq_gen(ugh_subreq_t *r, strp u_host);
int ugh_subreq_del(ugh_subreq_t *r, uint32_t ft_type, int ft_errno);
int ugh_subreq_del_after_module(ugh_subreq_t *r);

strp ugh_subreq_get_host(ugh_subreq_t *r);
in_port_t ugh_subreq_get_port(ugh_subreq_t *r);

#define UGH_UPSTREAM_FT_OFF             0x0000
#define UGH_UPSTREAM_FT_ERROR           0x0001
#define UGH_UPSTREAM_FT_TIMEOUT         0x0002
#define UGH_UPSTREAM_FT_INVALID_HEADER  0x0004
#define UGH_UPSTREAM_FT_HTTP_500        0x0008
#define UGH_UPSTREAM_FT_HTTP_502        0x0010
#define UGH_UPSTREAM_FT_HTTP_503        0x0020
#define UGH_UPSTREAM_FT_HTTP_504        0x0040
#define UGH_UPSTREAM_FT_HTTP_404        0x0080
#define UGH_UPSTREAM_FT_HTTP_5XX        0x0100
#define UGH_UPSTREAM_FT_HTTP_4XX        0x0200
#define UGH_UPSTREAM_FT_TIMEOUT_CONNECT 0x0400

void ugh_subreq_wait(ugh_client_t *c);

/* headers received from backend */
ugh_header_t *ugh_subreq_header_get_nt(ugh_subreq_t *r, const char *data);
ugh_header_t *ugh_subreq_header_get(ugh_subreq_t *r, const char *data, size_t size);
ugh_header_t *ugh_subreq_header_set(ugh_subreq_t *r, const char *data, size_t size, char *value_data, size_t value_size);

/* headers to send to backend */
ugh_header_t *ugh_subreq_header_out_get_nt(ugh_subreq_t *r, const char *data);
ugh_header_t *ugh_subreq_header_out_get(ugh_subreq_t *r, const char *data, size_t size);
ugh_header_t *ugh_subreq_header_out_set(ugh_subreq_t *r, const char *data, size_t size, char *value_data, size_t value_size);

/* ### parser ### */

int ugh_parser_client(ugh_client_t *c, char *data, size_t size);
int ugh_parser_client_body(ugh_client_t *c, char *data, size_t size);
int ugh_parser_subreq(ugh_subreq_t *r, char *data, size_t size);
int ugh_parser_chunks(ugh_subreq_t *r, char *data, size_t size);

/* ### channel ### */

typedef struct ugh_channel_message
	ugh_channel_message_t;

struct ugh_channel_message
{
	unsigned tag;
	strt content_type;
	strt body;
};

/* channel of proxy type is removed, when all planned subrequests are finished
 * and all responses were sent to at least one client */

#define UGH_CHANNEL_LONG_POLL 0
#define UGH_CHANNEL_INTERVAL_POLL 1

#define UGH_CHANNEL_WORKING 0
#define UGH_CHANNEL_DELETED 1

#define UGH_CHANNEL_PERMANENT 0
#define UGH_CHANNEL_PROXY 1

struct ugh_channel
{
	ugh_server_t *s;
	aux_pool_t *pool;

	ev_async wev_message;
	ev_timer wev_timeout;
	ev_tstamp timeout;

	void *messages_hash; /* JudyL (etag -> ugh_channel_message_t *) */
	/* TODO last-modified / if-modified-since mechanism */

	void *clients_hash; /* Judy1 (ugh_client_t *) */ /* XXX Judy is not required here */
	size_t clients_size; /* XXX we can use plain array with this size variable */

	void *subreqs_hash; /* Judy1 (ugh_subreq_t *) */

	strt channel_id;

	unsigned status:1; /* WORKING|DELETED */
	unsigned type:1; /* PERMANENT|PROXY */
};

ugh_channel_t *ugh_channel_add(ugh_server_t *s, strp channel_id, unsigned type, ev_tstamp timeout);
ugh_channel_t *ugh_channel_get(ugh_server_t *s, strp channel_id);
int ugh_channel_del(ugh_server_t *s, strp channel_id);

int ugh_channel_add_subreq(ugh_channel_t *ch, ugh_subreq_t *s);

int ugh_channel_add_message(ugh_channel_t *ch, strp body, strp content_type, ugh_subreq_t *r); /* subrequest is optional parameter here */
int ugh_channel_get_message(ugh_channel_t *ch, ugh_client_t *c, ugh_channel_message_t **m, unsigned type, Word_t *etag);

/* ### status ### */

#define UGH_HTTP_OK                                 0x00 /* 200 */
#define UGH_HTTP_CREATED                            0x01 /* 201 */
#define UGH_HTTP_ACCEPTED                           0x02 /* 202 */
#define UGH_HTTP_NON_AUTHORITATIVE_INFORMATION      0x03 /* 203 */
#define UGH_HTTP_NO_CONTENT                         0x04 /* 204 */
#define UGH_HTTP_RESET_CONTENT                      0x05 /* 205 */
#define UGH_HTTP_PARTIAL_CONTENT                    0x06 /* 206 */
#define UGH_HTTP_MULTI_STATUS                       0x07 /* 207 */
#define UGH_HTTP_MULTIPLE_CHOICES                   0x08 /* 300 */
#define UGH_HTTP_MOVED_PERMANENTLY                  0x09 /* 301 */
#define UGH_HTTP_MOVED_TEMPORARILY                  0x0A /* 302 */
#define UGH_HTTP_SEE_OTHER                          0x0B /* 303 */
#define UGH_HTTP_NOT_MODIFIED                       0x0C /* 304 */
#define UGH_HTTP_USE_PROXY                          0x0D /* 305 */
#define UGH_HTTP_UNUSED                             0x0E /* 306 */
#define UGH_HTTP_TEMPORARY_REDIRECT                 0x0F /* 307 */
#define UGH_HTTP_BAD_REQUEST                        0x10 /* 400 */
#define UGH_HTTP_UNAUTHORIZED                       0x11 /* 401 */
#define UGH_HTTP_PAYMENT_REQUIRED                   0x12 /* 402 */
#define UGH_HTTP_FORBIDDEN                          0x13 /* 403 */
#define UGH_HTTP_NOT_FOUND                          0x14 /* 404 */
#define UGH_HTTP_NOT_ALLOWED                        0x15 /* 405 */
#define UGH_HTTP_NOT_ACCEPTABLE                     0x16 /* 406 */
#define UGH_HTTP_PROXY_AUTHENTICATION_REQUIRED      0x17 /* 407 */
#define UGH_HTTP_REQUEST_TIME_OUT                   0x18 /* 408 */
#define UGH_HTTP_CONFLICT                           0x19 /* 409 */
#define UGH_HTTP_GONE                               0x1A /* 410 */
#define UGH_HTTP_LENGTH_REQUIRED                    0x1B /* 411 */
#define UGH_HTTP_PRECONDITION_FAILED                0x1C /* 412 */
#define UGH_HTTP_REQUEST_ENTITY_TOO_LARGE           0x1D /* 413 */
#define UGH_HTTP_REQUEST_URI_TOO_LARGE              0x1E /* 414 */
#define UGH_HTTP_UNSUPPORTED_MEDIA_TYPE             0x1F /* 415 */
#define UGH_HTTP_REQUESTED_RANGE_NOT_SATISFIABLE    0x20 /* 416 */
#define UGH_HTTP_EXPECTATION_FAILED                 0x21 /* 417 */
#define UGH_HTTP_INTERNAL_SERVER_ERROR              0x22 /* 500 */
#define UGH_HTTP_METHOD_NOT_IMPLEMENTED             0x23 /* 501 */
#define UGH_HTTP_BAD_GATEWAY                        0x24 /* 502 */
#define UGH_HTTP_SERVICE_TEMPORARILY_UNAVAILABLE    0x25 /* 503 */
#define UGH_HTTP_GATEWAY_TIME_OUT                   0x26 /* 504 */
#define UGH_HTTP_HTTP_VERSION_NOT_SUPPORTED         0x27 /* 505 */
#define UGH_HTTP_VARIANT_ALSO_NEGOTIATES            0x28 /* 506 */
#define UGH_HTTP_INSUFFICIENT_STORAGE               0x29 /* 507 */
#define UGH_HTTP_TOO_MANY_REQUESTS                  0x2A /* 429 */
#define UGH_HTTP_UNAVAILABLE_FOR_LEGAL_REASONS      0x2B /* 451 */
#define UGH_HTTP_STATUS_MAX                         0x2C

extern const char *ugh_status_header [UGH_HTTP_STATUS_MAX];

/* ### module ### */

#if 1
typedef struct ugh_module_handle
	ugh_module_handle_t;

struct ugh_module_handle
{
	ugh_module_t *module;
	void *config;
	int (*handle) (ugh_client_t *c, void *data, strp body);
};

#define	UGH_MODULE_HANDLES_MAX 1024

extern ugh_module_handle_t ugh_module_handles [UGH_MODULE_HANDLES_MAX];
extern size_t ugh_module_handles_size;

#define ugh_module_handle_add(_module, _config, _handle) \
	ugh_module_handles[ugh_module_handles_size].module = &_module; \
	ugh_module_handles[ugh_module_handles_size].config = _config; \
	ugh_module_handles[ugh_module_handles_size].handle = _handle; \
	ugh_module_handles_size++

int ugh_module_handle_all(ugh_client_t *c);

/* XXX remove this macro */
#define ugh_module_config_get_last() ugh_module_handles[ugh_module_handles_size - 1].config
#endif

#if 1
extern coro_context ctx_main;
extern unsigned char is_main_coro;
#endif

/* ### variables ### */

typedef struct ugh_variable
	ugh_variable_t;

struct ugh_variable
{
	strt name;
	strp (*handle)(ugh_client_t *c, const char *name, size_t size, void *data);
	void *data;
};

void ugh_idx_variable(ugh_config_t *cfg, const char *name, size_t size);
ugh_variable_t *ugh_set_variable(ugh_config_t *cfg, const char *name, size_t size, ugh_variable_t *var);
ugh_variable_t *ugh_get_variable(ugh_config_t *cfg, const char *name, size_t size);

strp ugh_get_varvalue(ugh_client_t *c, const char *name, size_t size);

#define ugh_get_varvalue_nt(c, name) ugh_get_varvalue((c), (name), strlen(name))

/* ### ugh_template ### */

#define UGH_TEMPLATE_MAX_CHUNKS 32
#define UGH_TEMPLATE_MAX_LENGTH 4096

typedef struct ugh_template
	ugh_template_t;

struct ugh_template
{
	strp chunks;
	size_t chunks_size;
};

int  ugh_template_compile(ugh_template_t *t, char *data, size_t size, ugh_config_t *cfg);
strp ugh_template_execute(ugh_template_t *t, ugh_client_t *c);

#ifdef __cplusplus
}
#endif

#endif /* __UGH_H__ */
